mpitests/test_expval.py

```diff
@@ -22,99 +22,236 @@
 
-        O1 = qml.PauliY(wires=[0])
-        O2 = qml.PauliY(wires=[1])
 
-        dev.apply(
-            [qml.RX(theta, wires=[0]), qml.RX(phi, wires=[1]), qml.CNOT(wires=[0, 1])],
-            rotations=[*O1.diagonalizing_gates(), *O2.diagonalizing_gates()],
-        )
+def apply_operation_gates_qnode_nonparam(tol, dev_mpi, operation, Wires):
+    """Wrapper applying a non-parametric gate with QNode function."""
+    num_wires = numQubits
+    comm = MPI.COMM_WORLD
+    commSize = comm.Get_size()
+    num_global_wires = commSize.bit_length() - 1
+    num_local_wires = num_wires - num_global_wires
 
-        res = np.array([dev.expval(O1), dev.expval(O2)])
-        assert np.allclose(res, np.array([0, -np.cos(theta) * np.sin(phi)]), tol)
+    c_dtype = dev_mpi.c_dtype
 
-    def test_hadamard_expectation(self, theta, phi, tol):
-        """Test that Hadamard expectation value is correct"""
-        dev = qml.device(device_name, mpi=True, wires=3)
+    expected_output_cpu = np.zeros(1 << num_wires).astype(c_dtype)
+    local_state_vector = np.zeros(1 << num_local_wires).astype(c_dtype)
+    local_expected_output_cpu = np.zeros(1 << num_local_wires).astype(c_dtype)
 
-        O1 = qml.Hadamard(wires=[0])
-        O2 = qml.Hadamard(wires=[1])
+    state_vector = create_random_init_state(num_wires, dev_mpi.c_dtype)
+    comm.Bcast(state_vector, root=0)
 
-        dev.apply(
-            [qml.RY(theta, wires=[0]), qml.RY(phi, wires=[1]), qml.CNOT(wires=[0, 1])],
-            rotations=[*O1.diagonalizing_gates(), *O2.diagonalizing_gates()],
-        )
+    comm.Scatter(state_vector, local_state_vector, root=0)
+    dev_cpu = qml.device("lightning.qubit", wires=num_wires, c_dtype=c_dtype)
+
+    def circuit():
+        qml.StatePrep(state_vector, wires=range(num_wires))
+        operation(wires=Wires)
+        return qml.state()
 
-        res = np.array([dev.expval(O1), dev.expval(O2)])
-        expected = np.array(
-            [
-                np.sin(theta) * np.sin(phi) + np.cos(theta),
-                np.cos(theta) * np.cos(phi) + np.sin(phi),
-            ]
-        ) / np.sqrt(2)
-        assert np.allclose(res, expected, tol)
+    cpu_qnode = qml.QNode(circuit, dev_cpu)
+    expected_output_cpu = cpu_qnode().astype(c_dtype)
+    comm.Scatter(expected_output_cpu, local_expected_output_cpu, root=0)
 
+    mpi_qnode = qml.QNode(circuit, dev_mpi)
+    local_state_vector = mpi_qnode()
+
+    assert np.allclose(local_state_vector, local_expected_output_cpu, atol=tol, rtol=0)
+
+
+class TestExpval:
+    """Tests that expectation values are properly calculated or that the proper errors are raised."""
+
+    @pytest.mark.parametrize("C_DTYPE", [np.complex128, np.complex64])
+    @pytest.mark.parametrize(
+        "operation",
+        [
+            qml.PauliX,
+            qml.PauliY,
+            qml.PauliZ,
+            qml.Hadamard,
+            # qml.Identity,
```

Comment:
can we use pytest.mark.skipif to skip qml.Identity here?

Reply:
Done!